### PR TITLE
Fix RoomDirectory to join by alias whenever possible.

### PR DESCRIPTION
**Pull request by dbkr**
This doesn't really attempt to unravel the different code paths that joining a room can take (apart from commenting it a little). It just fixes this bug for now.

### CODE REVIEW

**Comment by dbkr** [comment]
wait, there's a better way to do this

**Comment by dbkr** [comment]
ok, all yours

File: src/components/structures/RoomDirectory.js [hunk 1]
```diff
@@ -81,20 +81,30 @@ module.exports = React.createClass({
...
         // extract the metadata from the publicRooms structure to pass
         // as out-of-band data to view_room, because we get information
         // here that we can't get other than by joining the room in some
         // cases.
         var room;
-        for (var i = 0; i < this.state.publicRooms.length; ++i) {
-            if (this.state.publicRooms[i].room_id == roomId) {
-                room = this.state.publicRooms[i];
-                break;
+        if (roomIdOrAlias[0] == '!') {
```
Comment (richvdh) [inline src/components/structures/RoomDirectory.js]: I'm probably missing something here, but doesn't this condition prevent the whole function working for aliases?

Comment (richvdh) [inline src/components/structures/RoomDirectory.js]: wait, I'm a crank

**Comment by richvdh** [comment]
ugh. everything is awful. but ok.

Are you sure this fixes #1532? I don't think the RoomDirectory component is involved in that.

**Comment by richvdh** [comment]
Not merging this right now because I don't want to babysit its deployment to /develop

**Comment by dbkr** [comment]
Ah, I made this less awful... and then somehow managed to not actually push the commit. Now requires https://github.com/matrix-org/matrix-react-sdk/pull/304 which fixes #1532. ptal :)

File: src/components/structures/RoomDirectory.js [hunk 2]
```diff
@@ -114,11 +116,21 @@ module.exports = React.createClass({
...
-        dis.dispatch({
-            action: 'view_room',
-            room_id: roomId,
+        var payload = {
             oob_data: oob_data,
-        });
+        };
+        if (roomAlias) {
+            payload.action = 'view_room_alias';
+            payload.room_alias = roomAlias;
```
Comment (richvdh) [inline src/components/structures/RoomDirectory.js]: Is it a mistake to throw away the room id if we know it? I'm worried about cases where the directory somehow gets out of sync with the `room_aliases` in the rooms, so you actually end up joining a different room to the one you chose in the directory.

Perhaps at least a comment to this effect?

Comment (richvdh) [inline src/components/structures/RoomDirectory.js]: Actually, shouldn't this logic of 'prefer the alias if we have it' be in MatrixChat?

**Comment by richvdh** [comment]
Certainly an improvement. Looks good modulo my comment.

**Comment by dbkr** [comment]
ptal

**Comment by richvdh** [comment]
lgtm

**Pull request merged**